forge/src/main/java/me/lucko/luckperms/forge/listeners/ForgeConnectionListener.java

```diff
@@ -0,0 +1,161 @@
+import me.lucko.luckperms.common.plugin.util.AbstractConnectionListener;
+import me.lucko.luckperms.forge.ForgeSenderFactory;
+import me.lucko.luckperms.forge.LPForgePlugin;
+import me.lucko.luckperms.forge.event.ConnectionEvent;
+import net.kyori.adventure.text.Component;
+import net.minecraft.Util;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraftforge.event.entity.player.PlayerEvent;
+import net.minecraftforge.eventbus.api.EventPriority;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ForgeConnectionListener extends AbstractConnectionListener {
+    private final LPForgePlugin plugin;
+    private final Map<UUID, CompletableFuture<Boolean>> pendingConnections;
+
+    public ForgeConnectionListener(LPForgePlugin plugin) {
+        super(plugin);
+        this.plugin = plugin;
+        this.pendingConnections = new ConcurrentHashMap<>();
+    }
+
+    @SubscribeEvent
+    public void onAuth(ConnectionEvent.Auth event) {
+        GameProfile profile = event.getProfile();
+
+        if (this.plugin.getConfiguration().get(ConfigKeys.DEBUG_LOGINS)) {
+            this.plugin.getLogger().info("Processing pre-login for " + profile.getId() + " - " + profile.getName());
+        }
+
+        if (this.pendingConnections.containsKey(profile.getId())) {
+            return;
+        }
+
+        CompletableFuture<Boolean> future = CompletableFuture.supplyAsync(() -> {
+            /* Actually process the login for the connection.
+               We do this here to delay the login until the data is ready.
+               If the login gets cancelled later on, then this will be cleaned up.
+
+               This includes:
+               - loading uuid data
+               - loading permissions
+               - creating a user instance in the UserManager for this connection.
+               - setting up cached data. */
+            try {
+                User user = loadUser(profile.getId(), profile.getName());
+                recordConnection(profile.getId());
+                this.plugin.getEventDispatcher().dispatchPlayerLoginProcess(profile.getId(), profile.getName(), user);
+                return true;
+            } catch (Exception ex) {
+                this.plugin.getLogger().severe("Exception occurred whilst loading data for " + profile.getId() + " - " + profile.getName(), ex);
+                this.plugin.getEventDispatcher().dispatchPlayerLoginProcess(profile.getId(), profile.getName(), null);
+                return !this.plugin.getConfiguration().get(ConfigKeys.CANCEL_FAILED_LOGINS);
+            }
+        }, this.plugin.getBootstrap().getScheduler().async());
+
+        this.pendingConnections.put(profile.getId(), future);
+    }
+
+    @SubscribeEvent
+    public void onPlayerLoadFromFile(PlayerEvent.LoadFromFile event) {
```

Comment:
TODO Fix: This event gets fired BEFORE the `player.connection` field has been set and results in the catch logic failing.
This may unfortunately require another Mixin as Forge doesn't have any events that fit.